web/core/src/player.ts

```diff
@@ -5,264 +5,428 @@ import { processEpisodeData } from './transformers';
+      Object.assign(this.replay, rest);
+      if (Array.isArray(steps)) {
+        this.replay.steps = steps;
+      }
+      needsRender = true;
     }
 
-    private play() {
-        if (this.playing) return;
-        this.playing = true;
-        if (this.replay && this.step === this.replay.steps.length - 1) {
-            this.setStep(0);
-        }
-        this.tick();
-        this.renderControls();
+    // Update steps from 'setSteps'
+    if (event.data.setSteps && this.replay) {
+      this.replay.steps = event.data.setSteps;
+      needsRender = true;
     }
 
-    private pause() {
-        if (!this.playing) return;
-        this.playing = false;
-        this.renderControls();
+    // Overwrite replay object if a full 'replay' is provided
+    if (event.data.replay) {
+      this.replay = event.data.replay;
+      needsRender = true;
     }
 
-    private tick = () => {
-        if (!this.playing || !this.replay) return;
+    // After any replay update, save it to HMR state
+    if (needsRender && this.replay) {
+      updateHMRState('replay', this.replay); // Save to HMR
+    }
 
-        if (this.step >= this.replay.steps.length - 1) {
-            this.playing = false;
-            this.renderControls();
-            return;
-        }
+    // Update the current step
+    if (typeof event.data.step === 'number') {
+      this.step = event.data.step;
+      updateHMRState('step', this.step); // Save to HMR
+      needsRender = true;
+    }
 
-        setTimeout(() => {
-            this.setStep(this.step + 1);
-            this.tick();
-        }, this.speed);
-    };
+    // If any data was updated and we have a replay object, call setData.
+    if (needsRender && this.replay) {
+      this.setData(this.replay, this.agents);
+    }
+  };
 
-    private renderControls() {
-        if (!this.replay || !this.showControls) {
-            this.controls.style.display = 'none';
-            return;
-        }
+  private setData(replay: ReplayData, agents: any[] = [], options: { skipRender?: boolean } = {}) {
+    this.replay = replay;
+    this.agents = agents;
 
-        this.controls.style.display = 'flex';
-        const maxSteps = this.replay.steps.length - 1;
+    if (!this.mounted) {
+      this.adapter.mount(this.viewer, this.replay);
+      this.mounted = true;
+    }
 
-        // Update only what's necessary
-        const newIconHTML = this.getIconHTML(this.playing ? 'pause' : 'play');
-        if (this.playPauseIconPath.outerHTML !== newIconHTML) {
-            this.playPauseIconPath.outerHTML = newIconHTML;
-            this.playPauseIconPath = this.playPauseButton.querySelector('path')!;
-        }
+    // TODO(michaelaaron) - Turn this into something more reasonable.
+    if (
+      this?.replay?.steps &&
+      !(this?.replay?.steps as any)?.[0]?.stepType &&
+      this?.replay?.configuration?.game === 'repeated_poker'
+    ) {
+      this.replay.steps = processEpisodeData(this.replay, 'repeated_poker');
```

Comment:
double checking that this is here on purpose?

Reply:
ya it is needed for running the local replay mode as we need to process the data for the replayer for the steps to be correct. Ack it's terrible and I should come up with an actual solution soon.